### DISCUSSION TITLE
blobstore:  proxy/allocator volume cache continue to serve read/write when allocator and clustermgr network partioned ?

### DISCUSSION THREAD

**Discussion by jianwei1216** [cubefs/cubefs#3456] [Q&A] [answered]
我最近在研究 blobstore 的 proxy/allocator 模块，我发现一个问题（不确定是不是问题）

<img width="1193" alt="图片" src="https://github.com/user-attachments/assets/dc28be32-fbcf-48cb-a88b-1092dc077e49">

1. proxy/allocator 服务在启动时会向 clustermgr 申请 4 个 volume
- 区分第一次申请直接从 volumeAllocator.actives.allocatorVols[host_"allocator_clientIP"]
- 区分第二次申请 OperTypeAllocVolume，持久化到 tokenTbl 经过 raft 存储 KV 中
2. proxy/allocator 服务会每隔 40 秒钟向 clustermgr 续租所有申请到 volume
- proxy/allocator 会剔除掉 full 的 volume
- clustermgr 会进一步判断所要续租的 volume 是否满了，是否过期了，是否还健康
3. clustermgr 会周期性检查已分配出去的 volume 是否已经过期了，如果过期了，就剔除掉

a> clustermgr server 侧
OperTypeAllocVolume ==> OperTypeRetainVolume ==> OperTypeExpireVolume

b> proxy/allocator client 侧
OperTypeAllocVolume ==> OperTypeRetainVolume （缺少超时剔除）

lease 租约机制应该需要 client / server 2侧共同遵守，（我个人理解）
client 侧：申请==>续租==>超时检查（不再使用，目前还是会继续使用）
server 侧：分配==>续租==>超时检查（标记为空闲）

我们无法避免这样的网络分区：

1. access <==> proxy/allocator 网络互通
2. proxy/allocator <==> clustermgr 网络不通

问题1：
我在源码中没有看到 proxy/allocator client 侧在 vid.Token 超时后从自身缓存的 volume 中删除掉该 vid，定时器只有续租，针对续租失败的情况，分2种情况处理：
case-1:   proxy/allocator client <==> clustermgr server 通信成功，会剔除掉续租失败的 vid
case-2：proxy/allocator client <==> clustermgr server 通信失败，直接就返回 error 了，并没有检查是否 vid.Token 是否超时，也就没有剔除掉续租失败的 vid

在 access <==> proxy/allocator 之间分配 vid 时，只检查了 vid 的容量和健康评分，并没有检测 vid.Token 是否超时，这也就意味着 proxy/allocator client 续租失败的 vid 依旧再服务于读写，而 clustermgr server 已经将该 vid.Token 判定超时，并标记为 idle 了

这个会不会引发其他问题？还是有意弱化续租失败的场景，尽可能让 proxy/allocator 服务于读写？

问题2: lease租约机制一般用于独占，blobstore期望proxy/allocator独占vid吗？

@Cloudstriff @pengtianyue025 @sejust @true1064 感谢各位朋友能够答疑解惑

我做了一个测试模拟 clustermgr <==> proxy/allcator 之间断开：
1.  pkill -9 clustermgr
2. 查看 proxy/allocator 的日志：报错向 clustermgr 续租失败
`2024/07/31 13:59:23.634550 [ERROR] blobstore/proxy/allocator/retain_volume_task.go:205 [3d4b88b893c89c61:687ec6b398a5b5cb] retain volume from clusterMgr failed: Post "http://127.0.0.1:9998/volume/retain": dial tcp 127.0.0.1:9998: connect: connection refused`
3. 查看 proxy/allocator 缓存的 volume: 依旧能够看到 allocator 中的 volume 没有失效
```
BS $> proxy allocator list --code_mode 11
List on proxy:http://127.0.0.1:9600 codemode:11 volumes:4 top:10

hosts:
 [
    {
        "Host": "http://127.0.0.1:8899",
        "N": 24,
        "Vids": [
            14,
            14,
            14,
            14,
            14,
            14,
            15,
            15,
            15,
            15,
            15,
            15,
            16,
            16,
            16,
            16,
            16,
            16,
            17,
            17,
            17,
            17,
            17,
            17
        ]
    }
]

disks:
 [
    {
        "Disk": 7,
        "N": 1,
        "Vids": [
            14
        ]
    },
    {
        "Disk": 1,
        "N": 2,
        "Vids": [
            15,
            17
        ]
    },
    {
        "Disk": 8,
        "N": 2,
        "Vids": [
            14,
            16
        ]
    },
    {
        "Disk": 5,
        "N": 3,
        "Vids": [
            15,
            16,
            17
        ]
    },
    {
        "Disk": 4,
        "N": 4,
        "Vids": [
            14,
            15,
            16,
            17
        ]
    },
    {
        "Disk": 6,
        "N": 4,
        "Vids": [
            14,
            15,
            16,
            17
        ]
    },
    {
        "Disk": 3,
        "N": 4,
        "Vids": [
            14,
            15,
            16,
            17
        ]
    },
    {
        "Disk": 2,
        "N": 4,
        "Vids": [
            14,
            15,
            16,
            17
        ]
    }
]
```
4. access 上传对象依旧可以成功
```
BS $> access put -f file-4M.txt -v
to upload size:4194304 hash:b(0)
2024/07/31 15:15:14.977812 [DEBUG] blobstore/common/rpc/lb.go:257 [271695d0f4db42c0-cmd:330a4f907cabd659] lb.doCtx: start /put?size=4194304&hashes=0
Reading: [==================================================] 100%
2024/07/31 15:15:15.118202 [DEBUG] blobstore/common/rpc/lb.go:324 [271695d0f4db42c0-cmd:330a4f907cabd659] lb.doCtx: the last host of request, try times: 1, code: 200, err: <nil>, host: http://localhost:9500

put location json    : {"cluster_id":1,"code_mode":11,"size":4194304,"blob_size":4194304,"crc":3201676738,"blobs":[{"min_bid":119,"vid":17,"count":1}]}
put location readable:
	Crc        : 3201676738   (0xbed5b5c2)
	ClusterID  : 1
	CodeMode   : 11           (EC3P3)
	Size       : 4194304      (4.0 MiB)
	BlobSize   : 4194304      (4.0 MiB)
	Blobs: (1) [
	 >:  0| MinBid: 119                  Vid: 17         Count: 1
	]
	--> Encode: 18 of 37 bytes
	--> Hex   : bed5b5c2010b808080028080800201771101
	--> Base64: vtW1wgELgICAAoCAgAIBdxEB
```

**Comment by Cloudstriff** [comment, marked as answer]
Question 1: We intentionally aim to mitigate the issue where the cluster may become unable to write during a failure of ClusterMgr or in the event of a network partition. Therefore, even if the Proxy lease renewal times out and fails, the current Volume can still be retained to maintain normal writing for a period of time, thereby gradually enhancing the system's disaster recovery capability.

Question 2: In the design of Blobstore, it is not a strict requirement for a Volume to be exclusively owned. In the system design, even if a Volume is held and assigned for writing by multiple Proxies, it will not cause issues. At most, this may lead to the Volume filling up earlier and resulting in write failures, which can be resolved by requesting other Volumes for writing. Therefore, we do not strictly require that a Volume must be exclusively owned.

**Reply by jianwei1216** [reply]
有 3 个后台任务是需要独占 volume 的：balancer/disk_droper/munaul_migrater


balancer :  用于扩容或负载均衡，将 VolumeUnit 在各个 Disk 上均衡，以实现容量均衡和负载均衡；
disk_dropper ：用于缩容，将计划下线的 Disk 上的 VolumeUnit 迁移到其它 Disk 上；
munaul_migrater ：用于人工干预 VolumeUnit 在 Disk 上的分布，无论是缩容、扩容还是负载均衡，在算法不能正确分布的情况下，可以手动干预分布；


balancer
```
collectTaskLoop() ==> collectionTask() ==> selectDisks(mgr.cfg.MaxDiskFreeChunkCnt, mgr.cfg.MinDiskFreeChunkCnt)
                              ==> genOneBalanceTask(ctx, disk) ==> mgr.IMigrator.AddTask(ctx, task)
```

disk_dropper
```
collectTaskLoop() ==> collectTask() ==> acquireDropDisk() ==> genDiskDropTasks() 
                              ==> listMigratingVuid(DiskID) ==> listUnMigratedVuid(DiskID) ==> mgr.IMigrator.AddTask(ctx, &t)
```

MigrateMgr AddTask
```
AddTask() ==> mgr.prepareQueue.PushTask(task.TaskID, task)
```

MigrateMgr prepareTask
```
prepareTask() ==> GetVolumeInfo(Vid) ==> LockVolume(Vid)  //加锁独占 Volume
                       ==> AllocVunitSafe(src_vuid, dst_vuid)
```

BlobNode DoMigrate  
```
src_vuid ==> dst_vuid（可以通过 ec 修复）
```

MigrateMgr finishTask
```
finishTaskLoop() ==> finishTask() ==> UpdateVolume(Vid, src_vuid, dst_vuid) ==> ReleaseVolumeUnit(src_vuid)
                            ==> UnlockVolume(Vid) //解锁释放 Volume 的独占
```



LockVolume ==> MigrateVolumeUnit ==> UnlockVolume

这里锁定 Volume 的目的是为了 src_vuid ==> dst_vuid 数据迁移期间， volume 不再接受用户新的写入，
目的：volume 的 bid 列表起码不再新增新写，再修复期间，只有已有的 bid 或者要删除的 bid



VolumeMgr LockVolume
```
func (v *VolumeMgr) LockVolume(ctx context.Context, vid proto.Vid) error {
	if !vol.canLock() {
		vol.lock.RUnlock()
		span.Warnf("can't lock volume, volume %d, current status(%d)", vid, status)
		return apierrors.ErrLockNotAllow
	}
}


func (vol *volume) canLock() bool {
	return vol.getStatus() == proto.VolumeStatusIdle //这里要求 vol.Status 必须为 Idle
}
```

问题：

LockVolume 的所操作的 Volume，未必就是 Idle 的，因为 Proxy 与Clustermgr 网络分区续租 Volume 失败后，还会依旧继续服务新写

在这种场景下，access 与 clustermgr 网络分区后，proxy 与 clustermgr 也处于同样的网络分区
- clustermgr 最新持有的 volume (1,2,3,4,12,13)
- Proxy 持有的却是旧的 volume (1,2,3,4,5,6)
- access 持有的也是旧的 volume (1,2,3,4,5,6) 
   - 迁移过程中， access 写 5,6 还是成功的，但这部分新增的数据不会被 balancer/disk_droper 迁移走数据 (会不会有丢失数据的风险？）
   - 迁移完成后，access 写 5,6 应该会失败，会一直向 proxy/mq 投递修补事件

这种不一致的情况，会带来数据一致性的风险吗？

**Reply by jianwei1216** [reply]
> 核心问题: 后台任务在对volume/volumeunit重新映射后，proxy或者access没有知晓的情况下，会不会产生数据安全风险？

> 尤其是在较长时间的网络分区，balancer/disk_droper已经修改了volume下多个chunk的映射位置，在access看来已经写入old-chunk成功了，但这部分数据未得到迁移，直接切换到new-chunk，这部分业务认为写成功的数据会不会无法访问

@Cloudstriff 答复如下：

不会有数据安全风险，lock volume实际会执行数据节点对应chunk的禁写，禁写确认完成后才会进行数据迁移，确保数据出于静止状态

**Reply by jianwei1216** [reply]
> lock volume实际会执行数据节点对应chunk的禁写，禁写确认完成后才会进行数据迁移，确保数据出于静止状态

@Cloudstriff  BlobNode 禁止 chunk 写入的代码逻辑能帮忙点一下吗？
